### PR TITLE
Reusable Workflow Experiment: Upstreaming commit message checks

### PR DESCRIPTION
**Pull request by josegomezr**
- Bringing checks from os-autoinst/openQA#5343.
- Setting up reusable workflows [0].

[0]: https://docs.github.com/en/actions/using-workflows/reusing-workflows.

This is an experiment :lab_coat: 

First Error:

image

---

There's a successful schedule of the check: https://github.com/os-autoinst/os-autoinst-common/actions/runs/6626891884

---

Proposed Agreement:

- Define all checks as `base-%{check-name}.yml` into `.github/workflows`.
- Use them in `%{check-name}.yml` in `.github/workflows`.
- Make PR's in each derived repo so the workflow is used instead of re-defined. See the snippet below:

Reusable Workflow usage on downstream repos:

```yaml
---
name: 'Commit Message Check'

on:
  pull_request:
  push:
    branches:
      # we must not fix commit messages when they already reached master
      - '!master'

jobs:
  call-check-commit-message:
    uses: os-autoinst/os-autoinst-common./.github/workflows/base-commit-message-checker.yml@master
    secrets:
      accessToken: "${{ secrets.GITHUB_TOKEN }}"
```

### CODE REVIEW

**Review by okurz — CHANGES_REQUESTED** [review]
Please incorporate the changes from https://github.com/os-autoinst/os-autoinst-common/pull/29 . Idea looks nice

**Comment by josegomezr** [comment]
The expressions are already present here (rebased after #29 got merged).

It'd be great if we can come up with a list of general Checks to add them here at once and go then over the downstream repositories referring to the checks defined here.

**Comment by okurz** [comment]
> It'd be great if we can come up with a list of general Checks to add them here at once and go then over the downstream repositories referring to the checks defined here.

Well, I guess that's the idea of the PR and I like the idea of the PR or what do you mean?

**Comment by josegomezr** [comment]
> what do you mean?

What checks can be generalized so they apply to all repos?

Like the commit message one is a simple task, that's even agnostic of the language.

How about
- Perl/Python/JS/%lang% code style/lint checks?
- Spell Checks?
- %insert here a check%

?

**Comment by perlpunk** [comment]
There was an attempt to also put tools/tidy in here, but it's not that trivial because of the way os-autoinst-distri-opensuse is using this script: https://progress.opensuse.org/issues/110142
It's solvable, but we didn't have time.

File: .github/workflows/base-commit-message-checker.yml
Comment (perlpunk) [inline .github/workflows/base-commit-message-checker.yml:15]: There are still some typo fixes and lowercase/uppercase differences that weren't updated here

File: .perltidyrc
Comment (foursixnine) [inline .perltidyrc:5]: Lets stick to 160, we're not in the 80's anymore, screens are bigger and I don't expect people to do a proper review using their phones.

File: .perltidyrc
Comment (foursixnine) [inline .perltidyrc:17]: It's repeated

File: bin/perlcritic
Comment (foursixnine) [inline bin/perlcritic:13]: why the double dirname?

File: bin/tidyall
Comment (foursixnine) [inline bin/tidyall:20]: Maybe this can be formatted to make it pop up on GH workflows (https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#setting-an-error-message)

File: bin/perlcritic
Comment (josegomezr) [reply bin/perlcritic:13]: `readlink $0` will resolve the path of the specific file including symlinks.

that's `a/b/c/os-autoinst-common/bin/perlcritic` but we need what's on `a/b/c/os-autoinst-common/lib/perlcritic`.

First `dirname` deletes `perlcritic`, the second deletes `bin` and we end up with the desired path.

See:

```bash
dirname a/b/c/os-autoinst-common/bin/perlcritic
#> a/b/c/os-autoinst-common/bin
dirname a/b/c/os-autoinst-common/bin
#> a/b/c/os-autoinst-common
```

File: bin/tidyall
Comment (josegomezr) [reply bin/tidyall:20]: This is a local-only tool, `tools/tidy` is not run in any workflow that I can find, `perlcritic` does through `run_unit_tests.sh` and Makefile targets.

The idea is that is a transparent wrapper over `tidyall` so no special text processing is needed on ever divergent bash scripts.

File: .github/workflows/base-commit-message-checker.yml
Comment (josegomezr) [reply .github/workflows/base-commit-message-checker.yml]: Brought the changes from master here. Resolving.

File: .perltidyrc
Comment (josegomezr) [reply .perltidyrc]: Removed

**Comment by josegomezr** [comment]
> There was an attempt to also put tools/tidy in here, but it's not that trivial because of the way os-autoinst-distri-opensuse is using this script: https://progress.opensuse.org/issues/110142

In this implementation, how `Perl::Tidy` is provisioned (cpanfile or RPM) is irrelevant, as long as the correct version is installed (that is to say: perl knows about it) then `bin/tidyall` works as expected.

When it's not installed then:

```
Can't locate Perl/Tidy.pm in @INC (you may need to install the Perl::Tidy module) (@INC contains: /usr/lib/perl5/site_perl/5.26.1/x86_64-linux-thread-multi /usr/lib/perl5/site_perl/5.26.1 /usr/lib/perl5/vendor_perl/5.26.1/x86_64-linux-thread-multi /usr/lib/perl5/vendor_perl/5.26.1 /usr/lib/perl5/5.26.1/x86_64-linux-thread-multi /usr/lib/perl5/5.26.1 /usr/lib/perl5/site_perl) at bin/tidyall line 4.
BEGIN failed--compilation aborted at bin/tidyall line 4.
```

If a more graceful error is expected, maybe beginning with:

```perl
eval { require Perl::Tidy; };

if($@){
    die "Perl::Tidy is not installed";
}
```

Would allow to report more gracefully that perltidy is not installed? 

I lean towards avoiding it. The setup guides should get each developer to a point where all perl dependencies are installed properly.

File: bin/tidyall
Comment (perlpunk) [inline bin/tidyall:1]: `tidyall` is provides by the perl module Code::TidyAll, so I think creating a script with the same name is unfortunate.
Like I said, tidyall provides tidy tools for several languages/tools, including perltidy.

We use it mainly because it has a caching functionality, so it is faster to use locally.
In CI we are using https://github.com/os-autoinst/openQA/blob/master/t/00-tidy.t and https://github.com/os-autoinst/os-autoinst/blob/master/xt/00-tidy.t (and yes, it's unfortunate that they are different :)

File: .perltidyrc
Comment (perlpunk) [inline .perltidyrc:1]: See https://github.com/os-autoinst/os-autoinst/blob/master/xt/00-tidy.t vs. https://github.com/os-autoinst/openQA/blob/master/.perltidyrc

So we would have to decide about which config we want to use in the future.

File: .perltidyrc
Comment (perlpunk) [reply .perltidyrc:5]: Despite hearing that argument very often, the retricted line length is not about the screen width, but also about horizontal eye movement, and additionally side by side diff views are harder with longer lines.
But in my team I'm in the minority, most prefer longer lines anyway ;-)

**Comment by perlpunk** [comment]
Can you please make seperate pull requests for all these concerns? The sharing of the workflow was a good idea and seemsto work. The other things are a bit more complicated because some things are also used from os-autoinst-distri-opensuse.
Maybe you want to join our daily Team meeting tomorrow at 10:30 to discuss things?

File: bin/tidyall
Comment (josegomezr) [reply bin/tidyall:1]: I'm very open to suggestions on naming, the goal with this is to minimize scattering across repos and differences.

from your examples I see two ways of approaching linter checks:

- Run it as a test in `t/` (which I guess `prove` runs).
- Run it as a standalone command (which is what the `openQA` version is doing).

I simplified the latter, as I see there are may versions of the same script. But I don't know if either of the versions is more or less beneficial.

I guess it's also team agreement on which strategy to pick.

IMHO: I lean for the standalone command because it looks more similar to how other stacks work:

+ in Ruby you'd install dependencies & `bundle exec rubocop`
+ in Python you'd install dependencies, virtualenv [with your favourite virtualenv tool] & `{black|flake8|pylint}`.
+ in JS you install dependencies & `eslint`.

File: .perltidyrc
Comment (josegomezr) [reply .perltidyrc:1]: Looks like `tidyall_ok` uses the default config file:

https://github.com/houseabsolute/perl-code-tidyall/blob/10425fb8e720c06f96d3935bc98b2955fc3e3294/lib/Test/Code/TidyAll.pm#L23-L29

looks like it'd be a matter of providing this one as a symlink and that would be it ?

Comment (perlpunk) [reply .perltidyrc:1]: Yeah, the point is, if we use the symlink to this one here, we would first need to decide which of the two different configs we want.
I think it's better if we do that first by providing single PRs to both repos with the suggested new config and look at the tidy changes, so we can decide.

File: bin/tidyall
Comment (perlpunk) [reply bin/tidyall:1]: But I'm also wondering about other things.
I use `tools/tidy -o` to only tidy changed files.
Also it has a check-only mode.

And tools for development shouldn't be in `bin/`, as this is usually the directory that contains scripts that are packaged.

**Review by okurz — CHANGES_REQUESTED** [review]
I like the overall approach very much but I don't think it's feasible to bring in all those rules at once. Let's please start with the common ground of git commit message check and see that this is applied in all relevant downstream projects first. Then we can extend to tidy+perlcritic, etc. You can keep this PR as is and split out the git commit message check for a start into a separate PR and then rebase here, ok?

**Comment by josegomezr** [comment]
> You can keep this PR as is and split out the git commit message check for a start into a separate PR and then rebase here, ok?

Done, took a bit of problems with my local copy but it's there.

There's one thing that doesn't look quite right: the workflow for commit needs an approval. That's a fairly trivial check that can run automatically. Any idea how to make that? Nothing in the YAML's says anything about manual workflows

**Comment by perlpunk** [comment]
For everyone who does their first contribution on a github repo, workflow runs need to be approved. This was introduced by Github a while ago to prevent crypto miners from using resources.
So once your first PR has been merged, workflows will run automatically.
I will approve it manually for now.

**Review by perlpunk — APPROVED** [review]

**Review by okurz — APPROVED** [review]

**Pull request merged**
